[CLI] [FR] Add new CLI command for generating Dockerfile (issue #6532)

anuarkaliyev23 opened the pull request:
<!-- 🚨 We recommend pull requests be filed from a non-master branch on a repository fork (e.g. <username>:fix-xxx). 🚨 -->

## Related Issues/PRs

<!--
Please reference any related feature requests, issues, or PRs here. For example, `#123`. To automatically close the referenced items when this PR is merged, please use a closing keyword (close, fix, or resolve). For example, `Close #123`. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue for more information.
-->
Close #6532 

## What changes are proposed in this pull request?

This pull request proposes changes in CLI commands: it adds new `generate-dockerfile` command for `models`. It the outputs dockerfile with model directory in directory specified by CLI parameter. Resulting Dockerfile upon building and running behaves exactly like `models build-docker` image

example:
```bash
mlflow models generate-dockerfile -m <artifact> -d <name of the output directory>
```

To achieve this many functions of `build_docker` were reused and slightly refactored

## How is this patch tested?

<!--
If you're unsure about what to test, where to add tests, or how to run tests, please feel free to ask. We'd be happy to help.
-->
I tested locally both `generate-dockerfile` and `build-docker` commands with seemingly expected behavior. Python is not my primary language, so I would be glad to hear any advise regarding writing adequate tests
- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

<!--
Please describe how you confirmed the proposed feature/bug-fix/change works here. For example, if you fixed an MLflow client API, you could attach the code that didn't work prior to the fix but works now, or if you added a new feature on MLflow UI, you could attach a video that demonstrates the feature.
-->

## Does this PR change the documentation?
I am not entirely sure. I can see that `generate-dockerfile` cli command is present in `--help`, but I am not sure if any other documentation should be fixed for this matter

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

## Release Notes

[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)
    
Implemented cli command to generate dockerfile identical to the one in `build-docker` command


### Is this a user-facing change?

- [] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add new CLI command for generating dockerfiles for model tracked and stored with mlflow

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [x] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Created custom function for generating dockerfile which will be used in cli command

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Added functions for CLI command and in backend.py to generate dockerfile

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Implemented cli command to generate dockerfile identical to the one in `build-docker` command

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

harupy commented on mlflow/models/cli.py:
```diff
@@ -35,15 +35,15 @@ def commands():
-    env_manager=None,
-    install_mlflow=False,
-    enable_mlserver=False,
+        model_uri,
```
@anuarkaliyev23 Thanks for the PR! Can you revert unrelated changes?

anuarkaliyev23 replied:
Yes, I'm sorry, my IDE seems to change indents and I didn't notice.
Fixed

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

reverting unrelated changes

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

dbczumar commented on mlflow/models/cli.py:
```diff
@@ -155,6 +155,35 @@ def prepare_env(
+    Generates a dockerfile whose default entrypoint serves an MLflow model at port 8080, using the
+    python_function flavor.
+
+    Essentially, it generates a dockerfile, which would've been the base for the docker image in `build_docker`
+    function
```
```suggestion
    This dockerfile defines an image that is equivalent to the one produced by ``mlflow models build-docker``
```

dbczumar commented on mlflow/models/cli.py:
```diff
@@ -155,6 +155,35 @@ def prepare_env(
+    """
+    env_manager = env_manager or _EnvManager.CONDA
+    backend = _get_flavor_backend(model_uri, docker_build=True, env_manager=env_manager)
+    backend.generate_dockerfile(
```
Can we check `backend.can_build_image()` and throw if the backend cannot build an image, which indicates that we cannot build a dockerfile for it? Can we add a similar check to https://github.com/mlflow/mlflow/blob/9ea22cdf4d18204e6b92585a028d40341667ac6e/mlflow/models/cli.py#L209?

anuarkaliyev23 replied:
Added in both cases :+1:

dbczumar commented on mlflow/pyfunc/backend.py:
```diff
@@ -282,23 +286,66 @@ def can_score_model(self):
             # Can not find conda
             return False
 
+    def generate_dockerfile(
+            self, model_uri, output_directory="generate_dockerfile_output", install_mlflow=False, mlflow_home=None, enable_mlserver=False
+    ):
+        copy_model_into_container = self.copy_model_into_container_wrapper(model_uri, install_mlflow, enable_mlserver)
+        pyfunc_entrypoint = _pyfunc_entrypoint(self._env_manager, model_uri, install_mlflow, enable_mlserver)
+
+        mlflow_home = os.path.abspath(mlflow_home) if mlflow_home else None
+
+        is_conda = self._env_manager == em.CONDA
+        setup_miniconda = SETUP_MINICONDA if is_conda else ""
+        setup_pyenv_and_virtualenv = "" if is_conda else SETUP_PYENV_AND_VIRTUALENV
+
+        os.makedirs(output_directory, exist_ok=True)
+
+        _logger.debug("Created all folders in path", extra={"output_directory": output_directory})
+        install_mlflow = _get_mlflow_install_step(output_directory, mlflow_home)
+
+        custom_setup_steps = copy_model_into_container(output_directory) if copy_model_into_container else ""
+        dockerfile_text = _generate_dockerfile_raw(
+            setup_miniconda=setup_miniconda,
+            setup_pyenv_and_virtualenv=setup_pyenv_and_virtualenv,
+            install_mlflow=install_mlflow,
+            custom_setup_steps=custom_setup_steps,
+            entrypoint=pyfunc_entrypoint
+        )
+        _logger.debug("generated dockerfile text", extra={"dockerfile": dockerfile_text})
+
+        with open(os.path.join(output_directory, "Dockerfile"), "w") as dockerfile:
+            dockerfile.write(dockerfile_text)
+
     def build_image(
```
While we're at it, can we add the `build_image` and `generate_dockerfile` base methods to the FlavorBackend base class (https://github.com/mlflow/mlflow/blob/9ea22cdf4d18204e6b92585a028d40341667ac6e/mlflow/models/flavor_backend.py#L4) with a default implementation of `raise NotImplementedError`?

anuarkaliyev23 replied:
Done

dbczumar reviewed (commented):
@anuarkaliyev23 This is looking great! Just left a few small comments - we should be all set to merge after they're addressed. Thank you for your contribution!

dbczumar commented on mlflow/models/cli.py:
```diff
@@ -155,6 +155,35 @@ def prepare_env(
 
+@commands.command("generate-dockerfile")
+@cli_args.MODEL_URI_BUILD_DOCKER
+@click.option("--output-directory",
+              "-d",
+              default="generate-dockerfile-output",
+              help="output directory for generating dockerfile")
```
Can we accept a path rather than a directory, e.g. `/tmp/my/custom/path/Dockerfile` and default to `Dockerfile` so that the default behavior produces a dockerfile in the current working directory?

harupy replied:
+1 for using a path. Can we validate the specified Dockerfile path doesn't exist?

anuarkaliyev23 replied:
I started with a path, but encountered problems with `model_dir`. At first, I thought, that function will generate a `Dockerfile` only, but since `model_dir` used to copy model into image --- it should be present also.
This led me to believe that even if I pass a desired path to `Dockerfile` it will create a `model_dir ` in a parent directory, which seems to me as a little unexpected\not obvious, which can result in confusion for user --- since it only accepts dockerfile path, user can assume that only dockerfile is needed to build an image and can move it somewhere else *without moving `model_dir`*. 
Maybe I'm a little too cautious here, but folder seems as not ideal, but at least an obvious option

As for validating if a path doesn't exist --- function ensures creation of a path with `os.makedirs`, which allows us to pass any path

I am open to discussions and suggestions --- maybe I am just being too cautious here?

dbczumar replied:
@anuarkaliyev23 This is a great point. I think that producing a directory is the best option given that wrinkle. Can we rename the default directory to `mlflow-dockerfile` and update the helpstring to indicate that the directory contains a `Dockerfile` and also the resolved model, if `model-uri` is specified?

dbczumar commented on mlflow/models/cli.py:
```diff
@@ -155,6 +155,35 @@ def prepare_env(
+@cli_args.MLFLOW_HOME
+@cli_args.INSTALL_MLFLOW
+@cli_args.ENABLE_MLSERVER
+def generate_dockerfile(model_uri, output_directory, env_manager, mlflow_home, install_mlflow, enable_mlserver):
```
Can we add a test for this command? I think that https://github.com/mlflow/mlflow/blob/master/tests/models/test_cli.py is a good suite to add to

anuarkaliyev23 replied:
I will add test once we can find consensus on directory parameter in function, since it can potentially change a function hence a test.
At the moment I'm just not sure if we might need to change a function to accept a path to Dockerfile instead of directory :sweat_smile: , so until you approve decision to leave directory as an argument or decide to change a function's aruments, I will hold on a test for a little bit :sweat_smile:

dbczumar replied:
Responded here: https://github.com/mlflow/mlflow/pull/6591/files#r956774884. I think that a directory output makes sense :)

anuarkaliyev23 replied:
Yep, @dbczumar, sorry for the delay --- we had country holidays up until Wednesday :)
I'm gonna try and update pull request today to address all remaining comments and issues

anuarkaliyev23 replied:
I have written a test to ensure that `model_dir` and `Dockerfile` are generated and `Dockerfile` is not empty
However, I am somewhat struggling with test launching --- I am running `dev/run-python-tests` but it fails in 3 cases for unrelated reasons (not installed `kubernetes`, `google.cloud` and `pyarrow` modules)
When I tried to purposefully fail tests they are also didn't appear as failed once I ran the script.

Could it be related to issue https://github.com/mlflow/mlflow/issues/1096?

anuarkaliyev23 replied:
@harupy @dbczumar  Should we add `kubernetes`, `google-cloud-storage` and `pyarrow` to `requirements/test_requirements.txt` to ensure valid launch of `dev/run-python-tests`?
Should it be a separate issue?

harupy commented on mlflow/pyfunc/backend.py:
```diff
@@ -282,23 +286,66 @@ def can_score_model(self):
+        _logger.debug("generated dockerfile text", extra={"dockerfile": dockerfile_text})
+
+        with open(os.path.join(output_directory, "Dockerfile"), "w") as dockerfile:
+            dockerfile.write(dockerfile_text)
+
     def build_image(
         self, model_uri, image_name, install_mlflow=False, mlflow_home=None, enable_mlserver=False
     ):
+        copy_model_into_container = self.copy_model_into_container_wrapper(model_uri, install_mlflow, enable_mlserver)
+        pyfunc_entrypoint = _pyfunc_entrypoint(self._env_manager, model_uri, install_mlflow, enable_mlserver)
+        _build_image(
+            image_name=image_name,
+            mlflow_home=mlflow_home,
+            env_manager=self._env_manager,
+            custom_setup_steps_hook=copy_model_into_container,
+            entrypoint=pyfunc_entrypoint,
+        )
+
+    def copy_model_into_container_wrapper(self, model_uri, install_mlflow, enable_mlserver):
         def copy_model_into_container(dockerfile_context_dir):
+            # This function have to be included in another, since `_build_image` function in `docker_utils` accepts only
+            # single-argument function like this
             model_cwd = os.path.join(dockerfile_context_dir, "model_dir")
-            os.mkdir(model_cwd)
+            pathlib.Path(model_cwd).mkdir(parents=True, exist_ok=True)
             if model_uri:
                 model_path = _download_artifact_from_uri(model_uri, output_path=model_cwd)
                 return """
                     COPY {model_dir} /opt/ml/model
                     RUN python -c \
                     'from mlflow.models.container import _install_pyfunc_deps;\
                     _install_pyfunc_deps(\
-                        "/opt/ml/model", \
-                        install_mlflow={install_mlflow}, \
-                        enable_mlserver={enable_mlserver}, \
-                        env_manager="{env_manager}")'
+                         "/opt/ml/model", \
+                         install_mlflow={install_mlflow}, \
+                         enable_mlserver={enable_mlserver}, \
+                         env_manager="{env_manager}")'
```
Can we revert this change?

anuarkaliyev23 committed:
[CLI] [FR] Add new CLI command for generating Dockerfile (issue #6532)

Changing documentation as suggested in review

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Adding checks if backend can build an image for `generating-dockerfile` and `build-docker` commands

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

harupy commented on mlflow/models/docker_utils.py:
```diff
@@ -152,6 +152,41 @@ def _get_mlflow_install_step(dockerfile_context_dir, mlflow_home):
         ).format(version=mlflow.version.VERSION, maven_proxy=maven_proxy)
 
 
+def _generate_dockerfile_raw(
```
```suggestion
def _generate_dockerfile_content(
```

Can we rename this function `_generate_dockerfile_content`?

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Adding methods `build_image` and `generate_dockerfile` as abstract methods to flavor_backend.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Reverting unnecessary indent

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Rename `generate_dockerfile_raw` to `generate_dockerfile_content` as was proposed during review

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Remove @abstractmethod decorators to pass lint check

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Update function documentation to clarify use of directory path

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

format using `black`

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Add test for generate-dockerfile

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

format with `black`

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Changing default parameter for generate-dockerfile folder to `mlflow-dockerfile`

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

harupy commented on tests/models/test_cli.py:
```diff
@@ -413,6 +414,28 @@ def test_prepare_env_fails(sk_model):
+    if enable_mlserver:
+        extra_args.append("--enable_mlserver")
+
+    output_directory_name = pyfunc_build_image(model_uri, extra_args=extra_args)
```
```suggestion
    output_directory_name = pyfunc_generate_dockerfile(model_uri, extra_args=extra_args)
```

Is this line correct? I think we should call `pyfunc_generate_dockerfile` here.

anuarkaliyev23 replied:
Yep, you are correct, my bad :sweat:

harupy commented on tests/helper_functions.py:
```diff
@@ -87,6 +86,31 @@ def score_model_in_sagemaker_docker_container(
     return _evaluate_scoring_proc(proc, 5000, data, content_type, activity_polling_timeout_seconds)
 
 
+def pyfunc_generate_dockerfile(model_uri=None, extra_args=None):
```
```suggestion
def pyfunc_generate_dockerfile(output_directory, model_uri=None, extra_args=None):
```

Can we add `output_directory` and pass a temporary directory created by pytest to this argument?

anuarkaliyev23 replied:
Yep, it will be more convenient to do so
Done

harupy commented on tests/helper_functions.py:
```diff
@@ -87,6 +86,31 @@ def score_model_in_sagemaker_docker_container(
+        cmd += ["--mlflow-home", mlflow_home]
+    if extra_args:
+        cmd += extra_args
+    p = subprocess.Popen(cmd)
+    assert p.wait() == 0, "Failed to generate dockerfile for model %s" % model_uri
```
```suggestion
    subprocess.run(cmd, check=True)
```

harupy commented on tests/models/test_cli.py:
```diff
@@ -413,6 +414,28 @@ def test_prepare_env_fails(sk_model):
+        extra_args.append("--enable_mlserver")
+
+    output_directory_name = pyfunc_build_image(model_uri, extra_args=extra_args)
+    assert Path(output_directory_name).is_dir()
+    assert Path(output_directory_name).joinpath("Dockerfile").exists()
+    assert Path(output_directory_name).joinpath("model_dir").is_dir()
+    # Assert file is not empty
+    assert Path(output_directory_name).joinpath("Dockerfile").stat().st_size != 0
```
```suggestion
    output_directory = Path(output_directory_name)
    assert output_directory.is_dir()
    assert output_directory.joinpath("Dockerfile").exists()
    assert output_directory.joinpath("model_dir").is_dir()
    # Assert file is not empty
    assert output_directory.joinpath("Dockerfile").stat().st_size != 0
```

nit

anuarkaliyev23 committed:
Update tests/models/test_cli.py

change function to pyfunc_generate_dockerfile in test

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
Update tests/helper_functions.py

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
Update tests/models/test_cli.py

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Changed pyfunc_generate_dockerfile function to accept directory and pyfunc passes a temporary folder to it

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

harupy commented on tests/models/test_cli.py:
```diff
@@ -413,6 +415,30 @@ def test_prepare_env_fails(sk_model):
 
 
+@pytest.mark.parametrize("enable_mlserver", [True, False])
+def test_generate_dockerfile(sk_model, enable_mlserver):
```
```suggestion
def test_generate_dockerfile(sk_model, enable_mlserver, tmp_path):
```

harupy commented on tests/models/test_cli.py:
```diff
@@ -413,6 +415,30 @@ def test_prepare_env_fails(sk_model):
+    if enable_mlserver:
+        extra_args.append("--enable_mlserver")
+
+    with tempfile.TemporaryDirectory() as temp_dir:
+        pyfunc_generate_dockerfile(temp_dir, model_uri, extra_args=extra_args)
+        output_directory = Path(temp_dir)
+        assert output_directory.is_dir()
+        assert output_directory.joinpath("Dockerfile").exists()
+        assert output_directory.joinpath("model_dir").is_dir()
+        # Assert file is not empty
+        assert output_directory.joinpath("Dockerfile").stat().st_size != 0
```
```suggestion
    pyfunc_generate_dockerfile(tmp_path, model_uri, extra_args=extra_args)
    output_directory = Path(temp_dir)
    assert output_directory.is_dir()
    assert output_directory.joinpath("Dockerfile").exists()
    assert output_directory.joinpath("model_dir").is_dir()
    # Assert file is not empty
    assert output_directory.joinpath("Dockerfile").stat().st_size != 0
```

anuarkaliyev23 replied:
@harupy 
I haven't found any fixtures for `tmp_path`
is it some default fixture in pyfunc?
I'm sorry if it is a stupid question -- I'm not experienced in Python at all

harupy replied:
> is it some default fixture in pyfunc?

Yes, it's provided by pytest.

harupy commented on mlflow/models/cli.py:
```diff
@@ -155,6 +155,46 @@ def prepare_env(
+@click.option(
+    "--output-directory",
+    "-d",
+    default="generate-dockerfile-output",
```
As @dbczumar suggested, can we change the default value to `mlflow-dockerfile`?

anuarkaliyev23 replied:
I'm sorry, it might've gotten lost somewhere in stash, I should've checked before resolving

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Change default option for output-directory for generating dockerfile to mlflow-dockerfile

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
Update tests/models/test_cli.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

anuarkaliyev23 committed:
Update tests/models/test_cli.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

harupy commented on tests/models/test_cli.py:
```diff
@@ -413,6 +415,29 @@ def test_prepare_env_fails(sk_model):
+        extra_args.append("--enable_mlserver")
+
+    pyfunc_generate_dockerfile(tmp_path, model_uri, extra_args=extra_args)
+    output_directory = Path(temp_dir)
```
Looks like `temp_dir` here is undefined.

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Fixing invalid `temp_dir` reference --> `tmp_path`

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 commented:
@harupy @dbczumar Should we add kubernetes, google-cloud-storage and pyarrow to requirements/test_requirements.txt to ensure valid launch of dev/run-python-tests?
Should it be a separate issue?

Duplicated as a general comment for easier discussion

harupy commented:
I think `pip install -e .[extra]` should install `kubernetes`, `google-cloud-storage` and `pyarrow`.

https://github.com/mlflow/mlflow/blob/c9f64728f93331a4b1d70d685c17694fb2ac45fc/setup.py#L130

harupy commented on tests/models/test_cli.py:
```diff
@@ -413,6 +415,29 @@ def test_prepare_env_fails(sk_model):
+        extra_args.append("--enable_mlserver")
+
+    pyfunc_generate_dockerfile(tmp_path, model_uri, extra_args=extra_args)
+    output_directory = Path(tmp_path)
```
```suggestion
```

Nit: we can remove this line. `tmp_path` is a `Path` instance.

anuarkaliyev23 committed:
Update tests/models/test_cli.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

harupy commented on tests/models/test_cli.py:
```diff
@@ -37,7 +39,7 @@
     get_safe_port,
     pyfunc_serve_and_score_model,
-    PROTOBUF_REQUIREMENT,
+    PROTOBUF_REQUIREMENT, pyfunc_generate_dockerfile,
```
```suggestion
    PROTOBUF_REQUIREMENT,
    pyfunc_generate_dockerfile,
```

to pass the lint check.

anuarkaliyev23 committed:
Update tests/models/test_cli.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

harupy commented on tests/models/test_cli.py:
```diff
@@ -413,6 +415,28 @@ def test_prepare_env_fails(sk_model):
+    if enable_mlserver:
+        extra_args.append("--enable_mlserver")
+
+    pyfunc_generate_dockerfile(tmp_path, model_uri, extra_args=extra_args)
+    assert output_directory.is_dir()
+    assert output_directory.joinpath("Dockerfile").exists()
+    assert output_directory.joinpath("model_dir").is_dir()
+    # Assert file is not empty
```
```suggestion
    output_directory = tmp_path.joinpath("output_directory")
    pyfunc_generate_dockerfile(output_directory, model_uri, extra_args=extra_args)
    assert output_directory.is_dir()
    assert output_directory.joinpath("Dockerfile").exists()
    assert output_directory.joinpath("model_dir").is_dir()
    # Assert file is not empty
```

anuarkaliyev23 committed:
Update tests/models/test_cli.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

harupy commented on tests/models/test_cli.py:
```diff
@@ -2,6 +2,8 @@
 import subprocess
 import sys
+import tempfile
+from pathlib import Path
```
```suggestion
```

Let's remove this. `Path` is no longer used.

harupy commented on tests/models/test_cli.py:
```diff
@@ -2,6 +2,8 @@
 import os
 import subprocess
 import sys
+import tempfile
```
```suggestion
```

anuarkaliyev23 committed:
Update tests/models/test_cli.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

anuarkaliyev23 committed:
Update tests/models/test_cli.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

harupy reviewed (approved):
Looks good to me once the lint errors in https://github.com/mlflow/mlflow/runs/8149563761?check_suite_focus=true are fixed :)

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

fixing typo in parameter

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 commented:
@harupy 
> Looks good to me once the lint errors in https://github.com/mlflow/mlflow/runs/8149563761?check_suite_focus=true are fixed :)

yep, also spotted one typo in tests

`pylint` seem to take on formatting differently than `black` in case of, for example, line length
Should I address line lengths in pylint if they are seem to be ok in `black` eyes?

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Removing unnecessary reimport of env-manager

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Removing unnecessary imports

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

harupy commented:
> Should I address line lengths in pylint errors if they are seem to be ok in black eyes?

Yes. Unfortunately, black doesn't help you format too-long lines.

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

lint format

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Explicitly override abstract methods `build_image` and `generate_dockerfile`

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Change ternary operator calling to simple if-else due to linting warnings

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Fix linting error W0126, W0125 --- passed function in if. Changed to valid argument

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 commented:
@harupy 
I believe I fixed lint errors
In the previous run there were some failed tests, but they seem to be unrelated to the feature
Tests for CLI were successful

harupy commented on mlflow/pyfunc/backend.py:
```diff
@@ -282,12 +289,79 @@ def can_score_model(self):
+        _logger.debug("Created all folders in path", extra={"output_directory": output_path})
+        install_mlflow = _get_mlflow_install_step(output_path, mlflow_home)
+
+        custom_setup_steps = ()
+        if model_uri:
+            custom_setup_steps = copy_model_into_container(output_path)
+        else:
+            custom_setup_steps = ""
```
```suggestion
        custom_setup_steps = copy_model_into_container(output_path)
```

Can we remove `if-else`?

anuarkaliyev23 committed:
Update mlflow/pyfunc/backend.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>

anuarkaliyev23 commented:
@dbczumar 
Can you please review these changes? :)

WallEnd commented:
@dbczumar, please help us review this PR. We are so waiting for these changes.

anuarkaliyev23 commented:
@harupy 
If @dbczumar is not available could we assign another member to review changes prior to merge?

dbczumar commented on mlflow/models/cli.py:
```diff
@@ -155,6 +155,46 @@ def prepare_env(
+    model_uri, output_directory, env_manager, mlflow_home, install_mlflow, enable_mlserver
+):
+    """
+    Generates a directory with dockerfile whose default entrypoint serves an MLflow model at port
+    8080, using the python_function flavor. Function accepts a directory path and not just a
+    dockerfile path because if model is specified --- it will generate additional folder with
+    model files inside aforementioned directory This dockerfile defines an image that is
+    equivalent to the one produced by ``mlflow models build-docker``
```
```suggestion
    Generates a directory with Dockerfile whose default entrypoint serves an MLflow model at port
    8080 using the python_function flavor. The generated Dockerfile is written to the specified output
    directory, along with the model (if specified). This Dockerfile defines an image that is
    equivalent to the one produced by ``mlflow models build-docker``
```

dbczumar commented on mlflow/models/cli.py:
```diff
@@ -155,6 +155,46 @@ def prepare_env(
+    "--output-directory",
+    "-d",
+    default="mlflow-dockerfile",
+    help="output directory for generating dockerfile",
```
```suggestion
    help="Output directory where the generated Dockerfile is stored",
```

dbczumar commented on mlflow/models/cli.py:
```diff
@@ -155,6 +155,46 @@ def prepare_env(
+    dockerfile path because if model is specified --- it will generate additional folder with
+    model files inside aforementioned directory This dockerfile defines an image that is
+    equivalent to the one produced by ``mlflow models build-docker``
+    """
```
Can we add a small `logger.info("Generating Dockerfile)` message here? If a model_uri is specified, can we include it in the message?

anuarkaliyev23 replied:
done

dbczumar commented on mlflow/models/cli.py:
```diff
@@ -155,6 +155,46 @@ def prepare_env(
+            mlflow_home=mlflow_home,
+            install_mlflow=install_mlflow,
+            enable_mlserver=enable_mlserver,
+        )
```
After generation completes successfully, can we add a small `logger.info("Generated Dockerfile in directory <directory_path>")` message here?

dbczumar reviewed (approved):
LGTM once two small cosmetic comments are addressed. Thanks @anuarkaliyev23 !

anuarkaliyev23 committed:
Update mlflow/models/cli.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>
Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
Update mlflow/models/cli.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>
Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

add logger messages prior to and after generating dockerfile

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Add model uri to logs

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

anuarkaliyev23 committed:
[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)

Fix long lines warning from linter

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

harupy commented on mlflow/models/cli.py:
```diff
@@ -155,6 +155,50 @@ def prepare_env(
+    if model_uri:
+        _logger.info("Generating Dockerfile for model %s", model_uri)
+    else:
+        _logger.info("Generating Dockerfile for unspecified model")
```
```suggestion
        _logger.info("Generating Dockerfile")
```

Do we need `for unspecified model`?

anuarkaliyev23 replied:
@dbczumar asked "Can we add a small logger.info("Generating Dockerfile) message here? If a model_uri is specified, can we include it in the message?"

Maybe I took too far "if model_uri is specified" part :sweat_smile: 

I think we can just add "Generating dockerfile for model <model_uri>" if it's OK to you?
Than it just will be `None` if model_uri was not specified

dbczumar replied:
@anuarkaliyev23 Your latest changes look great! Thanks!

anuarkaliyev23 committed:
Update mlflow/models/cli.py

Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>
Signed-off-by: Khambar Dussaliyev <anuarkaliyev23@gmail.com>
